src/internal/obj/local_client.go

```
@@ -1,126 +1,149 @@
 		c = &monkeyClient{c}
 	}
 	return newUniformClient(c), nil
 }
 
-type localClient struct {
-	root string
+type fsClient struct {
+	rootDir string
 }
 
-func (c *localClient) normPath(path string) string {
-	path = filepath.Clean(path)
-	return filepath.Join(c.root, path)
+func newFSClient(rootDir string) (Client, error) {
+	c := &fsClient{rootDir: filepath.Clean(rootDir)}
+	if c.rootDir == "" || c.rootDir == "/" || c.rootDir == "." {
+		panic("you probably didn't want to set the local client's root path to " + c.rootDir)
+	}
+	return c, c.init()
 }
 
-func (c *localClient) Put(_ context.Context, path string, r io.Reader) (retErr error) {
-	defer func() { retErr = c.transformError(retErr, path) }()
-	fullPath := c.normPath(path)
-
-	// Create the directory since it may not exist
-	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
+func (c *fsClient) Put(ctx context.Context, name string, r io.Reader) error {
+	staging := c.stagingPathFor(name)
+	final := c.finalPathFor(name)
+	f, err := os.Create(staging)
+	if err != nil {
 		return err
 	}
-	file, err := os.Create(fullPath)
-	if err != nil {
+	defer f.Close()
+	if _, err := io.Copy(f, r); err != nil {
 		return err
 	}
-	defer func() {
-		if err := file.Close(); retErr == nil {
-			retErr = err
-		}
-	}()
-	if _, err := io.Copy(file, r); err != nil {
+	if err := f.Close(); err != nil {
 		return err
 	}
-	return nil
+	return os.Rename(staging, final)
 }
 
-func (c *localClient) Get(_ context.Context, path string, w io.Writer) (retErr error) {
-	defer func() { retErr = c.transformError(retErr, path) }()
-	file, err := os.Open(c.normPath(path))
+func (c *fsClient) Get(ctx context.Context, name string, w io.Writer) (retErr error) {
+	defer func() { retErr = c.transformError(retErr, name) }()
+	f, err := os.Open(c.finalPathFor(name))
 	if err != nil {
 		return err
 	}
-	defer func() {
-		if err := file.Close(); retErr == nil {
-			retErr = err
-		}
-	}()
-	_, err = io.Copy(w, file)
+	defer f.Close()
```

Probably shouldn't discard the error.